src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseSpanBasedStringConcat.cs

```diff
@@ -0,0 +1,310 @@
+                        return;
+
+                    topMostConcatOperations.Add(topMostConcatOperation);
+                }
+
+                void ReportDiagnosticsOnRootConcatOperationsWithSubstringCalls(OperationBlockAnalysisContext context)
+                {
+                    //  We report diagnostics for all top-most concat operations that contain 
+                    //  direct or conditional substring invocations when there is an applicable span-based overload of
+                    //  the string.Concat method.
+                    //  We don't report when the concatenation contains anything other than strings or character literals.
+                    foreach (var operation in topMostConcatOperations)
+                    {
+                        if (ShouldBeReported(operation))
+                        {
+                            context.ReportDiagnostic(operation.CreateDiagnostic(Rule));
+                        }
+                    }
+
+                    topMostConcatOperations.Free(context.CancellationToken);
+                }
+            }
+
+            bool ShouldBeReported(IBinaryOperation topMostConcatOperation)
+            {
+                var concatOperands = FlattenBinaryOperation(topMostConcatOperation);
+
+                //  Bail if no suitable overload of 'string.Concat' exists.
+                if (!symbols.TryGetRoscharConcatMethodWithArity(concatOperands.Length, out _))
+                    return false;
+
+                bool anySubstringInvocations = false;
+                foreach (var operand in concatOperands)
+                {
+                    var value = WalkDownBuiltInImplicitConversionOnConcatOperand(operand);
+                    switch (value.Type?.SpecialType)
+                    {
+                        //  Report diagnostics only when operands are exclusively strings and character literals.
+                        case SpecialType.System_String:
+                        case SpecialType.System_Char when value is ILiteralOperation:
+                            if (IsAnyDirectOrConditionalSubstringInvocation(value))
+                                anySubstringInvocations = true;
+                            break;
+                        default:
+                            return false;
+                    }
+                }
+
+                return anySubstringInvocations;
+            }
+
+            bool IsAnyDirectOrConditionalSubstringInvocation(IOperation operation)
+            {
+                if (operation is IConditionalAccessOperation conditionallAccessOperation)
+                    operation = conditionallAccessOperation.WhenNotNull;
+
+                return operation is IInvocationOperation invocation && symbols.IsAnySubstringMethod(invocation.TargetMethod);
+            }
+        }
+
+        internal static ImmutableArray<IOperation> FlattenBinaryOperation(IBinaryOperation root)
```

Comment:
Might be much simpler to just write a specialized [OperationWalker](http://sourceroslyn.io/#Microsoft.CodeAnalysis/Operations/OperationWalker.cs,11) here.

Reply:
Okay, I'll reimplement this using an `OperationWalker`

Reply:
You're right. I've done that.